File: src/main/scala/com/exasol/cloudetl/helper/ExportParallelismCalculator.scala
package com.exasol.cloudetl.helper

import com.exasol.ExaMetadata
import com.exasol.cloudetl.parallelism.FixedUdfCountCalculator
import com.exasol.cloudetl.parallelism.MemoryUdfCountCalculator
import com.exasol.cloudetl.storage.StorageProperties

/**
 * A class that generates parallelism string for an export.
 */
final case class ExportParallelismCalculator(metadata: ExaMetadata, properties: StorageProperties) {

  /**
   * Returns a SQL string for parallel UDF instances when exporting data.
   *
   * It uses user provided value; if it is not provided, it uses calculated
   * cores per node in combination with {@code iproc()}.
   *
   * @return parallelism string
   */
  def getParallelism(): String =
    properties.getParallelism().getOrElse {
      val multiplier = getCoresPerNode()
      String.format("iproc(), mod(rownum,%d)", multiplier)
    }

  private[this] def getCoresPerNode(): Int = {
    val coresPerNode = Runtime.getRuntime().availableProcessors()
    math.min(coresPerNode, getCoresLimitedByMemory())
  }

  private[this] def getCoresLimitedByMemory(): Int =
    if (properties.hasUdfMemory()) {
      MemoryUdfCountCalculator(metadata, properties.getUdfMemory()).getNumberOfCoresLimitedByMemoryPerNode()
    } else {
      FixedUdfCountCalculator(metadata).getNumberOfCoresLimitedByMemoryPerNode()
    }

}

File: src/main/java/com/exasol/cloudetl/helper/ExportParallelismCalculator.java
package com.exasol.cloudetl.helper;

import com.exasol.ExaMetadata;
import com.exasol.cloudetl.parallelism.FixedUdfCountCalculator;
import com.exasol.cloudetl.parallelism.MemoryUdfCountCalculator;
import com.exasol.cloudetl.storage.StorageProperties;

/** Generates parallelism strings for exports. */
public final class ExportParallelismCalculator {
    private final ExaMetadata metadata;
    private final StorageProperties properties;

    /** Create a calculator. */
    public ExportParallelismCalculator(final ExaMetadata metadata, final StorageProperties properties) {
        this.metadata = metadata;
        this.properties = properties;
    }

    /** Factory for Java callers. */
    public static ExportParallelismCalculator apply(final ExaMetadata metadata, final StorageProperties properties) {
        return new ExportParallelismCalculator(metadata, properties);
    }

    /** @return export parallelism SQL fragment */
    public String getParallelism() {
        if (this.properties.getParallelism().isDefined()) {
            return this.properties.getParallelism().get();
        }
        final int multiplier = getCoresPerNode();
        return String.format("iproc(), mod(rownum,%d)", multiplier);
    }

    private int getCoresPerNode() {
        final int coresPerNode = Runtime.getRuntime().availableProcessors();
        return Math.min(coresPerNode, getCoresLimitedByMemory());
    }

    private int getCoresLimitedByMemory() {
        if (this.properties.hasUdfMemory()) {
            return new MemoryUdfCountCalculator(this.metadata, this.properties.getUdfMemory())
                    .getNumberOfCoresLimitedByMemoryPerNode();
        }
        return new FixedUdfCountCalculator(this.metadata).getNumberOfCoresLimitedByMemoryPerNode();
    }
}
